### PR TITLE
FFC param key validation / FFC libctx cleanups

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Dependant on PR #10909 (which is dependent on #10860 & #10897)

**Comment by slontis** [comment]
Moving out of WIP now that dependent PR's have been merged.

**Comment by slontis** [comment]
The 1st commit is the validation changes for DSA and DH.
The 2nd commit removes the libctx being propagated thru the dh/dsa methods by storing the libctx internally in the objects. (It was done in this PR to fix the check methods, but then needed to be cleaned up elsewhere also where it was used in the generation and compute methods).

**Comment by slontis** [comment]
rebased..

**Comment by slontis** [comment]
ping

**Comment by slontis** [comment]
travis error is just a timeout..

**Comment by slontis** [comment]
ping

File: crypto/dsa/dsa_check.c [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+int dsa_check_priv_key(const DSA *dsa, const BIGNUM *priv_key, int *ret)
+{
+    int ok = 0;
+    BIGNUM *two_powN = NULL, *upper;
+
+    *ret = 0;
+    two_powN = BN_new();
+    if (two_powN == NULL)
```
Comment (mattcaswell) [inline crypto/dsa/dsa_check.c]: This never seems to be used.

File: crypto/ffc/ffc_key_validate.c [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ */
+int ffc_validate_public_key_partial(const FFC_PARAMS *params,
+                                    const BIGNUM *pub_key, int *ret)
+{
+    int ok = 0;
+    BIGNUM *tmp = NULL;
+    BN_CTX *ctx = NULL;
+
+    *ret = 0;
+    ctx = BN_CTX_new();
```
Comment (mattcaswell) [inline crypto/ffc/ffc_key_validate.c]: Grrr. This should fail in linking. `BN_CTX_new()` is not available in the FIPS provider. You must use `BN_CTX_new_ex()`

File: crypto/ffc/ffc_key_validate.c [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+{
+    int ok = 0;
+    BIGNUM *tmp = NULL;
+    BN_CTX *ctx = NULL;
+
+    if (!ffc_validate_public_key_partial(params, pub_key, ret))
+        return 0;
+
+    if (params->q != NULL) {
+        ctx = BN_CTX_new();
```
Comment (mattcaswell) [inline crypto/ffc/ffc_key_validate.c]: And here

File: test/ffc_internal_test.c [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+    if (!TEST_false(ffc_generate_private_key_fips(ctx, params, N+1, 112, priv)))
+        goto err;
+    /* pass since 2s <= N <= len(q) */
+    if (!TEST_true(ffc_generate_private_key_fips(ctx, params, N, 112, priv)))
+        goto err;
+    /* pass since N = len(q) */
+    if (!TEST_true(ffc_validate_private_key(params->q, priv, &res)))
+        goto err;
+    /* pass since 2s <= N < len(q) */
+    if (!TEST_true(ffc_generate_private_key_fips(ctx, params, N/2, 112, priv)))
```
Comment (mattcaswell) [inline test/ffc_internal_test.c]: Nit: spaces around operators. Here and in `N+1` case above.

File: crypto/dsa/dsa_check.c (hunk 1 above)
Comment (slontis) [reply crypto/dsa/dsa_check.c]: Well spotted - copied from the dh_check.c which does need this for safe primes..

File: crypto/ffc/ffc_key_validate.c (hunk 2 above)
Comment (slontis) [reply crypto/ffc/ffc_key_validate.c]: Even more annoying is that it does not care about the libctx :)

File: crypto/ffc/ffc_key_validate.c (hunk 3 above)
Comment (slontis) [reply crypto/ffc/ffc_key_validate.c]: fixed

**Comment by slontis** [comment]
Updated to address code review comments.

**Comment by slontis** [comment]
Rebased

**Comment by slontis** [comment]
ping

**Comment by mattcaswell** [comment]
Oh...wait....travis failure appears relevant.

**Comment by slontis** [comment]
Who put that no config build in :)

**Comment by slontis** [comment]
Added  OPENSSL_NO_DSA and OPENSSL_NO_DH sections to the new test

**Comment by slontis** [comment]
ping

**Comment by slontis** [comment]
Merged to master.

**Pull request closed without merging**